Skip Mode (#755)
cylc/cylc-doc: src/user-guide/task-implementation/run-modes.rst

.. _task-run-modes:

Task Run Modes
==============

.. versionadded:: 8.4.0

   * Before 8.4.0 run modes could only be set for the workflow, not
     for individual tasks.
   * Skip mode.

As well as the default task implementation ("live" mode)
other implementations can be selected:

* By setting the scheduler run mode as an argument to the
  ``cylc play --mode`` command ("simulation" and "dummy" modes).
* By setting :cylc:conf:`[runtime][<namespace>]run mode` for
  "skip" mode.

#. :ref:`task-run-modes.simulation`: Intended to be used
   when developing a workflow, the scheduler simulates
   running a task without creating a "real" job.
#. :ref:`task-run-modes.dummy`: Intended to be used
   when developing a workflow, the scheduler replaces
   the workflow scripts with dummy scripts and returns
   to the default ("live") submission pathway.
#. :ref:`task-run-modes.skip`: Intended to allow more control of the workflow,
   the scheduler marks task outputs as completed.

.. note::

   The simulation and dummy workflow run modes ignore the
   :cylc:conf:`[runtime][<namespace>]run mode` setting.

.. _task-run-modes.dummy:

> Dummy Mode

Reviewer: I don't think dummy and simulation should have been moved to a "task run mode" page?

Author: Arguable, because they are still task modes, but I take your point, and will move them back.

.. _task-run-modes:

Task Run Modes
==============

.. versionadded:: 8.4.0

   * Before 8.4.0 run modes could only be set for the workflow, not
     for individual tasks.
   * Skip mode.

As well as the default task implementation ("live" mode)
other implementations can be selected:

* By setting the scheduler run mode as an argument to the
  ``cylc play --mode`` command ("simulation" and "dummy" modes).
* By setting :cylc:conf:`[runtime][<namespace>]run mode` for
  "skip" mode.

#. :ref:`task-run-modes.simulation`: Intended to be used
   when developing a workflow, the scheduler simulates
   running a task without creating a "real" job.
#. :ref:`task-run-modes.dummy`: Intended to be used

>    when developing a workflow, the scheduler replaces
>    the workflow scripts with dummy scripts and returns
>    to the default ("live") submission pathway.

Reviewer 2: The submission "pathway" thing again - I think that kinda refers to internal implementation and might confuse users. Could be much simpler, e.g.:
```suggestion
   when developing a workflow. Instead of the real configured task scripts,
   the scheduler submits local jobs with empty (do nothing) "script" items.
```

#. :ref:`task-run-modes.skip`: Intended to allow more control of the workflow,

>    the scheduler marks task outputs as completed.

Reviewer 2: ```suggestion
   the scheduler marks task outputs as completed without submitting a job.
```

.. note::

   The simulation and dummy workflow run modes ignore the
   :cylc:conf:`[runtime][<namespace>]run mode` setting.

.. _task-run-modes.dummy:

Dummy Mode
----------

**Dummy mode** replaces real jobs with background jobs on the
scheduler host which use ``sleep`` to simulate the run length according to
the settings described for simulation mode.
This avoids :term:`job runner` directives that request compute
resources for real workflow tasks, and it allows any workflow configuration to

> be run locally in dummy mode.

Reviewer 2: Partially redundant - should be combined with the description higher up?

Author: Deliberately somewhat similar - the other text is a top level abstract, this is the introduction to the documentation for dummy mode.

Author: Resolved: Fell out of use when I moved the workflow run modes back to the other file

Simulated Failure
^^^^^^^^^^^^^^^^^

Tasks always complete all custom outputs, by default they succeed.

   If you want to test individual pathways, use
   :ref:`skip mode <task-run-modes.skip>`.

You can set some or all instances of a task to fail using
:cylc:conf:`[runtime][<namespace>][simulation]fail cycle points`.
``fail cycle points`` takes either a list of cycle point strings or "all".

Tasks set to fail will succeed on their second or following simulated
submission. If you want all submissions to fail, set
:cylc:conf:`[runtime][<namespace>][simulation]fail try 1 only = False`.

For example, to simulate a task you know to be flaky on the half
hour but not on the hour:

.. code-block:: cylc

   [[get_observations]]
      execution retry delays = PT30S
      [[[simulation]]]
         fail cycle points = 2022-01-01T00:30Z,  2022-01-01T01:30Z

In another case you might not expect the retry to work, and want to test
whether your failure handling works correctly:

.. code-block:: cylc

   [[get_data]]
       execution retry delays = PT30S
       [[[simulation]]]
          fail try 1 only = false
          fail cycle points = 2022-01-01T03:00Z

Limitations
^^^^^^^^^^^

> Dummy mode can only be applied to a whole workflow.

Reviewer 2: ```suggestion
Simulation mode can only be applied to a whole workflow.
```
(Already said above for Dummy mode).

Alternate path branching is difficult to simulate effectively. You can
configure certain tasks to fail via
:cylc:conf:`[runtime][<namespace>][simulation]`, but all branches based
on mutually exclusive custom outputs will run because all custom outputs get
artificially completed in dummy mode and in simulation mode.

.. _task-run-modes.skip:

Skip Mode
---------

> Skip mode is designed as an aid to workflow control:

Reviewer 2: "it allows ..."

```suggestion
Skip mode is designed as an aid to workflow control. It:
```

Broadcast :cylc:conf:`[runtime][<namespace>]run mode` setting the
value to ``skip``.

Create a Graph Control Task
^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. admonition:: Scenario

   We have a large family to large family trigger.

   .. code-block:: cylc

      [task parameters]
          x = 1..N   # Where N is big.
      [scheduling]
          [[graph]]
              R1 = FAMILY:succeed-all => FAMILY2
      [runtime]
         [[FAMILY, FAMILY2]]
         [[task1<x>]]
            inherit = FAMILY
         [[task2<x>]]
            inherit = FAMILY2

   If we increase N, the number of dependencies being tracked
   by the scheduler is N^2.

Adding a skip mode task between the families:

.. code-block:: cylc

   FAMILY:succeed-all => skiptask => FAMILY2

Will reduce the number of dependencies to 2N.

Parameter Exclusion
^^^^^^^^^^^^^^^^^^^

.. admonition:: Scenario

   We want to skip a small number of tasks from a parameterized
   group of tasks:

   .. code-block:: cylc

      [task parameters]
          # House number 13 doesn't actually exist on this street...
          house_number = 1..20

We can use skip mode to make sure that a parameter task always

> succeeds without running anything:

Reviewer 2: Nice example :-)

.. versionadded:: 8.4.0

> <<<<<<< HEAD

Reviewer 3: :(

Author: Rah!
